aten/src/ATen/native/UnaryOps.cpp

```diff
@@ -384,6 +374,153 @@ Tensor& _clamp_min_out_cpu(Tensor& result, const Tensor& self, Scalar min) {
   return result;
 }
 
+Tensor clamp_with_tensors(const Tensor& self, const Tensor& min, const Tensor& max) {
+  Tensor result = at::empty({0}, self.options());
+  return at::clamp_with_tensors_out(result, self, min, max);
+}
+
+Tensor clamp_with_tensors_max(const Tensor& self, const Tensor& max) {
+  Tensor result = at::empty({0}, self.options());
+  return at::clamp_with_tensors_max_out(result, self, max);
+}
+
+Tensor clamp_with_tensors_min(const Tensor& self, const Tensor& min) {
+  Tensor result = at::empty({0}, self.options());
+  return at::clamp_with_tensors_min_out(result, self, min);
+}
+
+Tensor& clamp_with_tensors_(Tensor& self, const Tensor& min, const Tensor& max) {
+  return at::clamp_with_tensors_out(self, self, min, max);
+}
+
+Tensor& clamp_with_tensors_out(
+    Tensor& result,
+    const Tensor& self,
+    const Tensor& min,
+    const Tensor& max) {
+  TORCH_CHECK(!self.is_complex(), "clamp is not yet implemented for complex tensors.");
+  if (min.defined() && max.defined()) {
+    TORCH_CHECK(self.device().type() == DeviceType::CPU || self.device().type() == DeviceType::CUDA,
+                "clamp only supports CPU and CUDA device type, got: ", self.device().type());
+    TORCH_CHECK(self.layout() == Layout::Strided,
+                "clamp only supports strided layout, got: ", self.layout());
+
+    // We clamp max tensor to make sure that it is not smaller than min
+    // This is necessary to fix the divergent behaviour in AVX and base CPU kernels
+    // https://github.com/pytorch/pytorch/pull/32587#issuecomment-591389582
+    auto adjusted_max = at::clamp_with_tensors_min(max, min);
+
+    auto iter = TensorIterator();
+    iter.set_check_mem_overlap(true);
+    iter.add_output(result);
+    iter.add_input(self);
+    iter.add_input(min);
+    iter.add_input(adjusted_max);
+    iter.promote_common_dtype();
```

Comment:
How does promotion work in this case? If self is float but max or min are complex, does that mean I'm trying to clamp a complex tensor, which is disallowed? Should the check for complex be expanded to include the min and max tensors, too?

Reply:
Sorry, I've missing this case. Thanks for spotting it. I will add a check to include min and max tensors.